# 8th_homework

### Import Section

In [1]:
import numpy as np
import pandas as pd

from statsmodels.stats.weightstats import ztest
from statsmodels.stats.weightstats import ttest_ind

from scipy.stats import mannwhitneyu

c:\users\zsv\appdata\local\programs\python\python39\lib\site-packages\statsmodels\compat\pandas.py:65: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.
  from pandas import Int64Index as NumericIndex


### Global Settings Section

In [2]:
random_state_global = 0

## Theoretical Part

#### 1. Если мы рассчитали 95% доверительный интервал для среднего значения, то какие из следующих утверждений являются верными?

Выберите несколько вариантов и попытайтесь объяснить свой выбор:

1) Если многократно повторять эксперимент, то 95 % выборочных средних значений будут принадлежать рассчитанному нами доверительному интервалу.

2) Мы можем быть на 95% уверены, что среднее значение в генеральной совокупности принадлежит рассчитанному доверительному интервалу.

3) Если многократно повторять эксперимент, для каждой выборки рассчитывать свой доверительный интервал, то в 95 % случаев истинное среднее будет находиться внутри доверительного интервала.

4) Среднее значение в генеральной совокупности точно превышает нижнюю границу 95% доверительного интервала.

5) Среднее значение в генеральной совокупности точно принадлежит рассчитанному доверительному интервалу.

##### Вариант 3 выглядит правильным, так как 95% доверительный интервал - это 95% от всех построенных интервалов захватят истинное математическое ожидание генеральной совокупности. Аналогичным образом вариант 1 также кажется правильным.

#### 2. Если мы увеличиваем объем выборки в два раза (при условии, что показатель стандартного отклонения остается неизменным), то 95% доверительный интервал

Выберите один вариант из списка

1) стал более узким

2) возможны оба варианта

3) стал более широким

##### Вариант 1, доверительный интервал станет более узким, так как при его построении используется величина Standard Error, которая обратно пропорциональна объёму выборки.

#### 3. В центре 95% доверительного интервала, рассчитанного по выборочным значениям, находится:

Выберите один вариант из списка

1) Значение стандартной ошибки среднего

2) Выборочное среднее значение

3) Среднее значение генеральной совокупности

##### Вариант 2, выборочное среднее значние, так как это значение будет использоваться при построении интервала +- SE * критерий.

#### 4. Часто на практике нулевая гипотеза отклоняется, и различия считаются статистически достоверными, если p < 0,05. Однако часто в статистике используется более жесткий критерий достоверности различий, например, при условии, что p < 0,01. Значение p-уровня значимости, которое выбирается, в качестве порога обозначается буквой α (альфа). Например, если исследователь решил, что α = 0,05, то и нулевая гипотеза будет отклоняться при условии, что p < 0,05. 

#### Если в определенной ситуации весьма рискованно отклонить нулевую гипотезу, когда она на самом деле верна, то лучше использовать показатель α равный 

Выберите один вариант из списка

1) 0,1

2) 0,001

3) 0,05

4) 0,5

##### Вариант 2, 0,001. Данное значение крайне мало по сравнению с общепринятым уровнем статистической значимости 0,05. Однако данное значение отражает вероятность совершения ошибки 1 рода, ошибочное отвержение нулевой гипотезы, поэтому в данном случае его необходимо минимизировать.

#### 5. Данные некоторого исследования сообщают нам, что средний рост детей в 14 лет составляет 166 сантиметров. Однако это лишь выборочная оценка, и исследователи рассчитали 95% доверительный интервал, который составил [160, 172]. Укажите верные утверждения:

Выберите несколько вариантов и попытайтесь объяснить свой выбор:

1) У нас достаточно оснований отклонить нулевую гипотезу, что среднее в генеральной совокупности равняется 173

2) Вероятность того, что истинное среднее значение больше 172, составляет 0,01

3) У нас достаточно оснований отклонить нулевую гипотезу, что среднее в генеральной совокупности равняется 158.

4) Доверительный интервал не может иметь такие границы, т. к. выборочное стандартное отклонение равняется 10, следовательно 
доверительный интервал должен быть значительно шире.

##### Утверждение 1 верно, мы можем с 95% уверенностью утверждать, что мат. ожидание генеральной совокупности отлично от 173.

##### Утверждение 2 ложно, вероятность будет 0.025.

##### Утверждение 3 верно, аналогично первому утверждению.

##### Утверждение 4 ложно, так как доверительный интервал зависит не только от выборочного станжартного отклонения, но и от объёма выборки.

#### 6. Предположим, нулевой гипотезой вашего исследования являлось предположение, что конверсия в генеральной совокупности равняется 0.4. Вы получили p = 0,12 и не смогли отклонить нулевую гипотезу. Однако позже выяснилось, что конверсия в генеральной совокупности действиетльно равна 0.4. Как можно оценить результаты?

Какой вариант корректный по вашему мнению и почему?

1) Вы не совершали ни ошибку первого рода, ни ошибку второго рода.

2) Вы совершили ошибку первого рода

3) Вы совершили ошибку второго рода

##### Вариант 1 верный, так как мы приняли нулевую гипотезу и позже выяснилось, что предполагаемая конверсия оказалась равна действительной.

## Practical Part

In [3]:
df_data =pd.read_csv('ab_datas.csv')

In [4]:
df_data.head()

user_id                   timestamp      group landing_page  converted  \
0   851104  2017-01-21 22:11:48.556739    control     old_page          0   
1   804228  2017-01-12 08:01:45.159739    control     old_page          0   
2   661590  2017-01-11 16:55:06.154213  treatment     new_page          0   
3   853541  2017-01-08 18:28:03.143765  treatment     new_page          0   
4   864975  2017-01-21 01:52:26.210827    control     old_page          1   

   revenue  
0        0  
1        0  
2      100  
3        0  
4      250

In [5]:
df_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294478 entries, 0 to 294477
Data columns (total 6 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       294478 non-null  int64 
 1   timestamp     294478 non-null  object
 2   group         294478 non-null  object
 3   landing_page  294478 non-null  object
 4   converted     294478 non-null  int64 
 5   revenue       294478 non-null  int64 
dtypes: int64(3), object(3)
memory usage: 13.5+ MB


In [6]:
df_data.isna().sum()

user_id         0
timestamp       0
group           0
landing_page    0
converted       0
revenue         0
dtype: int64

In [7]:
df_data.describe()

user_id      converted        revenue
count  294478.000000  294478.000000  294478.000000
mean   787974.124733       0.119659     101.967719
std     91210.823776       0.324563     103.602744
min    630000.000000       0.000000       0.000000
25%    709032.250000       0.000000       0.000000
50%    787933.500000       0.000000     100.000000
75%    866911.750000       0.000000     250.000000
max    945999.000000       1.000000     250.000000

In [8]:
df_data['revenue'].value_counts()

0      123000
250     85863
100     85615
Name: revenue, dtype: int64

In [9]:
df_data['landing_page'].value_counts()

old_page    147239
new_page    147239
Name: landing_page, dtype: int64

**Выводы:**

Данные не содержат пропусков и аномалий, предобработка не требуется.

Далее будем рассматривать только исследуемый признак "revenue" с разбиением на две группы по признаку "landing_page". Признак "landing_page" важнее чем "group", так как отражает фактическое разбиение клиентов на две группы. Также данный признак разделяет клиентов на две равные группы.

In [10]:
revenue_a = df_data.loc[df_data['landing_page'] == 'old_page', ['revenue']].copy()
revenue_b = df_data.loc[df_data['landing_page'] == 'new_page', ['revenue']].copy()

del df_data

In [11]:
revenue_a.describe()

revenue
count  147239.000000
mean       87.855459
std       102.506361
min         0.000000
25%         0.000000
50%       100.000000
75%       250.000000
max       250.000000

In [12]:
revenue_b.describe()

revenue
count  147239.000000
mean      116.079979
std       102.767999
min         0.000000
25%         0.000000
50%       100.000000
75%       250.000000
max       250.000000

Основные статистики обеих выборок схожи за исключением средних арифметических. Таким образом, предварительный анализ выявил различия в двух выборках, однако необходимо провести проверки статистических гипотез для более определённого ответа.

Далее сформулируем нулевую и альтернативную гипотезы, установим уровень статистической значимости альфа = 0.05.

Н0: средний чек в обеих выборках не имеет статистически значимых отличий.

Н1: средний чек имеет статистически значимые отличия.

Сперва выполним Z-test.

In [13]:
ztest(revenue_a, revenue_b)

(array([-74.61358756]), array([0.]))

P-value получилось равным нулю, что сигнализирует о значительных отличиях выборок.

Чтобы убедиться, что тест выполнен корректно, попробуем оценить суммы выборок.

In [14]:
print(f'''
Сумма значений контрольной выборки:\t\t{np.sum(revenue_a)[0]}
Сумма значений экспериментальной выборки:\t{np.sum(revenue_b)[0]}
Отношение сумм значений выборок:\t\t{(np.sum(revenue_b)[0] / np.sum(revenue_a)[0]):.4}
''')


Сумма значений контрольной выборки:		12935750
Сумма значений экспериментальной выборки:	17091500
Отношение сумм значений выборок:		1.321



Выборки действительно сильно отличаются.

In [15]:
ttest_ind(revenue_a, revenue_b)

(array([-74.61358756]), array([0.]), 294476.0)

In [16]:
mannwhitneyu(revenue_a, revenue_b)

MannwhitneyuResult(statistic=array([9.08703273e+09]), pvalue=array([0.]))

Т-тест и критерий Манна-Уитни также выявили значительные отличия, p-value = 0.

Таким образом, Н0 гипотеза отвергается в пользу Н1: выборки имеют статистически значимые отличия. Рассмотрев основные статистики выборок, можно установить, что средний чек экспериментальной выборки увеличился более чем на 32%.